src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -0,0 +1,286 @@
+            for (let lz = 0; lz < t.params.groupSize.z; lz++) {
+              for (let ly = 0; ly < t.params.groupSize.y; ly++) {
+                for (let lx = 0; lx < t.params.groupSize.x; lx++) {
+                  const groupIndex = (gz * t.params.numGroups.y + gy) * t.params.numGroups.x + gx;
+                  const localIndex = (lz * t.params.groupSize.y + ly) * t.params.groupSize.x + lx;
+                  const globalIndex = groupIndex * invocationsPerGroup + localIndex;
+                  const expected = getBuiltinValue(
+                    { x: gx, y: gy, z: gz },
+                    { x: lx, y: ly, z: lz }
+                  );
+                  if (output[globalIndex * 4 + 0] !== expected.x) {
+                    return new Error(
+                      `${name}.x failed at group(${gx},${gy},${gz}) local(${lx},${ly},${lz}))\n` +
+                        `    expected: ${expected.x}\n` +
+                        `    got:      ${output[globalIndex * 4 + 0]}`
+                    );
+                  }
+                  if (output[globalIndex * 4 + 1] !== expected.y) {
+                    return new Error(
+                      `${name}.y failed at group(${gx},${gy},${gz}) local(${lx},${ly},${lz}))\n` +
+                        `    expected: ${expected.y}\n` +
+                        `    got:      ${output[globalIndex * 4 + 1]}`
+                    );
+                  }
+                  if (output[globalIndex * 4 + 2] !== expected.z) {
+                    return new Error(
+                      `${name}.z failed at group(${gx},${gy},${gz}) local(${lx},${ly},${lz}))\n` +
+                        `    expected: ${expected.z}\n` +
+                        `    got:      ${output[globalIndex * 4 + 2]}`
+                    );
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+      return undefined;
+    };
+
+    // Check [[builtin(local_invocation_index)]] values.
+    t.expectGPUBufferValuesEqual(
+      localIndexBuffer,
+      new Uint32Array([...iterRange(totalInvocations, x => x % invocationsPerGroup)])
+    );
+
+    // Check [[builtin(local_invocation_id)]] values.
+    const checkLocalId = (outputData: Uint32Array) => {
+      return checkEachIndex(outputData, 'local_invocation_id', (_, localId) => localId);
+    };
+    t.expectGPUBufferValuesPassCheck(localIdBuffer, checkLocalId, {
+      type: Uint32Array,
+      typedLength: totalInvocations * 4,
+    });
+
+    // Check [[builtin(global_invocation_id)]] values.
+    const getGlobalId = (groupId: vec3, localId: vec3) => {
+      return {
+        x: groupId.x * t.params.groupSize.x + localId.x,
+        y: groupId.y * t.params.groupSize.y + localId.y,
+        z: groupId.z * t.params.groupSize.z + localId.z,
+      };
+    };
+    const checkGlobalId = (outputData: Uint32Array) => {
+      return checkEachIndex(outputData, 'global_invocation_id', getGlobalId);
```

Comment:
These can be made slightly less unwieldy with type inference (removing `: Uint32Array`) and lambda-style functions (removing `return`). Pushed a commit for this since it's hard to explain.

Reply:
Thanks, looks much better now.